### PR TITLE
Exodus payload caching

### PR DESCRIPTION
**Pull request by thebevrishot**
- Remove StyleCop from Test projects
- Add Encode function to Transaction Encoder
- Add Exodus payload to database

closes https://github.com/zcoinofficial/ztm/issues/161 https://github.com/zcoinofficial/ztm/issues/155

### CODE REVIEW

File: src/Ztm.Zcoin.NBitcoin/Exodus/ITransactionEncoder.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@ namespace Ztm.Zcoin.NBitcoin.Exodus
 {
     public interface ITransactionEncoder
     {
+        byte[] Encode(ExodusTransaction transaction);
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/ITransactionEncoder.cs]: Method in interface should ordered by ascending.

File: src/Ztm.Zcoin.NBitcoin/Exodus/TransactionEncoder.cs [hunk 2]
```diff
@@ -24,6 +24,34 @@ public TransactionEncoder(IEnumerable<ITransactionPayloadEncoder> encoders)
             this.encoders = encoders.ToDictionary(e => e.Type);
         }
 
+        public byte[] Encode(ExodusTransaction transaction)
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/TransactionEncoder.cs]: Same as previous issue.

File: src/Ztm.Zcoin.NBitcoin/Exodus/TransactionEncoder.cs [hunk 3]
```diff
@@ -24,6 +24,34 @@ public TransactionEncoder(IEnumerable<ITransactionPayloadEncoder> encoders)
...
+            using (var writer = new BinaryWriter(stream, Encoding.UTF8))
+            {
+                var type = transaction.Type;
+
+                writer.Write(IPAddress.HostToNetworkOrder((short)transaction.Version));
+                writer.Write(IPAddress.HostToNetworkOrder((short)transaction.Type));
+
+                if (!this.encoders.TryGetValue(type, out var encoder))
+                {
+                    throw new TransactionFieldException(
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/TransactionEncoder.cs]: We should use `ArgumentException` instead. `TransactionFieldException` was designed to catch invalid exodus payload.

File: src/Ztm.Zcoin.NBitcoin/Exodus/ITransactionPayloadEncoder.cs [hunk 4]
```diff
@@ -7,6 +7,7 @@ public interface ITransactionPayloadEncoder
     {
         int Type { get; }
 
+        void Encode(BinaryWriter writer, ExodusTransaction transaction);
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/ITransactionPayloadEncoder.cs]: Same as previous issue.

File: src/Ztm.Zcoin.NBitcoin/Exodus/ExodusTransaction.cs [hunk 5]
```diff
@@ -21,6 +21,8 @@ protected ExodusTransaction(BitcoinAddress sender, BitcoinAddress receiver)
 
         public BitcoinAddress Sender { get; }
 
+        public abstract int Type { get; }
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/ExodusTransaction.cs]: This is a duplicated with `Id`.

File: src/Ztm.Zcoin.NBitcoin/Exodus/TransactionPayloadEncoder.cs [hunk 6]
```diff
@@ -13,12 +13,34 @@ protected TransactionPayloadEncoder()
 
         public abstract int Type { get; }
 
+        protected abstract void Encode(BinaryWriter writer, ExodusTransaction transaction);
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/TransactionPayloadEncoder.cs]: Same as previous issue.

File: src/Ztm.Zcoin.NBitcoin/Exodus/TransactionPayloadEncoder.cs [hunk 7]
```diff
@@ -48,6 +70,21 @@ protected static PropertyId DecodePropertyId(BinaryReader reader)
             return new PropertyId(value);
         }
 
+        void ITransactionPayloadEncoder.Encode(BinaryWriter writer, ExodusTransaction transaction)
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/TransactionPayloadEncoder.cs]: Same as previous issue.

File: src/Ztm.Zcoin.NBitcoin/Exodus/TransactionPayloadEncoder.cs [hunk 8]
```diff
@@ -13,12 +13,34 @@ protected TransactionPayloadEncoder()
...
         protected abstract ExodusTransaction Decode(
             BitcoinAddress sender,
             BitcoinAddress receiver,
             BinaryReader payload,
             int version);
 
+        /// <summary>
+        /// Write a <see cref="PropertyAmount"/> to <paramref name="writer"/>.
+        /// </summary>
+        protected static void EncodePropertyAmount(BinaryWriter writer, PropertyAmount amount)
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/TransactionPayloadEncoder.cs]: Same as previous issue.

File: src/Ztm.Zcoin.NBitcoin/Exodus/TransactionPayloadEncoder.cs [hunk 9]
```diff
@@ -13,12 +13,34 @@ protected TransactionPayloadEncoder()
...
+        {
+            var value = IPAddress.HostToNetworkOrder(amount.Indivisible);
+
+            writer.Write(value);
+        }
+
+        /// <summary>
+        /// Write a <see cref="PropertyId"/> to <paramref name="writer"/>.
+        /// </summary>
+        protected static void EncodePropertyId(BinaryWriter writer, PropertyId id)
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/TransactionPayloadEncoder.cs]: Same here.

File: src/Ztm.Zcoin.NBitcoin/Exodus/SimpleSendEncoder.cs [hunk 10]
```diff
@@ -8,6 +8,26 @@ public sealed class SimpleSendEncoder : TransactionPayloadEncoder
     {
         public override int Type => SimpleSendV0.StaticId;
 
+        protected override void Encode(BinaryWriter writer, ExodusTransaction transaction)
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/SimpleSendEncoder.cs]: Same as previous issue.

File: src/Ztm.Zcoin.NBitcoin/Exodus/SimpleSendEncoder.cs [hunk 11]
```diff
@@ -8,6 +8,26 @@ public sealed class SimpleSendEncoder : TransactionPayloadEncoder
...
+        {
+            if (!(transaction is SimpleSendV0))
+            {
+                throw new InvalidOperationException("The transaction type is not simple send v0.");
+            }
+
+            var simpleSend = (SimpleSendV0)transaction;
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/SimpleSendEncoder.cs]: Better to use `as` operator so we can check if it not simple send with null check.

File: src/Ztm.Zcoin.NBitcoin/Exodus/SimpleSendEncoder.cs [hunk 12]
```diff
@@ -8,6 +8,26 @@ public sealed class SimpleSendEncoder : TransactionPayloadEncoder
     {
         public override int Type => SimpleSendV0.StaticId;
 
+        protected override void Encode(BinaryWriter writer, ExodusTransaction transaction)
+        {
+            if (!(transaction is SimpleSendV0))
+            {
+                throw new InvalidOperationException("The transaction type is not simple send v0.");
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/SimpleSendEncoder.cs]: We should use `ArgumentException` instead.

Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/Exodus/SimpleSendEncoder.cs]: Message is quite incorrect due to this class support any version of simple send.

File: src/Ztm.Data.Entity/Contexts/Main/ExodusPayload.cs [hunk 13]
```diff
@@ -0,0 +1,12 @@
+using NBitcoin;
+
+namespace Ztm.Data.Entity.Contexts.Main
+{
+    public sealed class ExodusPayload
+    {
+        public uint256 TransactionHash { get; set; }
+        public Script Sender { get; set; }
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/ExodusPayload.cs]: We don't need `Sender` to be `Script` here due to Exodus only working with the address, not script.

File: src/Ztm.Data.Entity/Contexts/Main/ExodusPayload.cs [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+        public Script Receiver { get; set; }
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/ExodusPayload.cs]: Same here.

File: src/Ztm.Data.Entity/Contexts/MainDatabase.cs [hunk 15]
```diff
@@ -142,6 +154,11 @@ protected virtual void ConfigureTransaction(EntityTypeBuilder<Transaction> build
             builder.Property(e => e.LockTime).IsRequired();
 
             builder.HasKey(e => e.Hash);
+            builder.HasOne(e => e.ExodusPayload)
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/MainDatabase.cs]: `Transaction > ExodusPayload` does not look right. We should reverse the relationship instead (e.g `ExodusPayload > Transaction`).

File: src/Ztm.Data.Entity/Contexts/Main/Transaction.cs [hunk 16]
```diff
@@ -19,5 +19,7 @@ public Transaction()
         public SortedSet<BlockTransaction> Blocks { get; set; }
         public SortedSet<Input> Inputs { get; set; }
         public SortedSet<Output> Outputs { get; set; }
+
+        public ExodusPayload ExodusPayload { get; set; }
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/Transaction.cs]: This one should be grouped together with others navigation properties.

File: src/Ztm.Data.Entity/Contexts/Main/ExodusPayload.cs (hunk 13 above)
Comment (thebevrishot) [reply src/Ztm.Data.Entity/Contexts/Main/ExodusPayload.cs]: Change it to string and store bitcoin address instead

File: src/Ztm.Zcoin.Synchronization/BlocksStorage.cs [hunk 17]
```diff
@@ -122,13 +122,23 @@ public async Task<Block> GetAsync(int height, CancellationToken cancellationToke
...
+                if (data == null)
                 {
                     return null;
                 }
 
-                data = rows[0];
-                previous = (rows.Length > 1) ? rows[1] : null;
+                if (data.Height == 0 && previous != null)
+                {
+                    return null;
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.Synchronization/BlocksStorage.cs]: We should throw with `InvalidOperationException` instead due to this one should not happened.

File: src/Ztm.Zcoin.Synchronization/BlocksStorage.cs [hunk 18]
```diff
@@ -122,13 +122,23 @@ public async Task<Block> GetAsync(int height, CancellationToken cancellationToke
...
+                }
+
+                if (data.Height != 0 && previous == null)
+                {
+                    return null;
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.Synchronization/BlocksStorage.cs]: NANI!!

File: src/Ztm.Zcoin.Synchronization/BlocksStorage.cs [hunk 19]
```diff
@@ -130,14 +130,9 @@ public async Task<Block> GetAsync(int height, CancellationToken cancellationToke
                     return null;
                 }
 
-                if (data.Height == 0 && previous != null)
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.Synchronization/BlocksStorage.cs]: We should keep this logic and throw when it hit so we know the data is in corrupted state.

**Pull request merged**